### ISSUE TITLE
Incorporate traversability measures into repo

### ISSUE THREAD

**Issue by schrum2** [schrum2/MarioDiffusion#164]
As mentioned in Issue #148, we need to make sure that when we have weird level slices, we are actually training on traversable level segments. It seems that any level segment from the real game data should be traversable by default, but it is possible that we are slicing scenes in ways that break things, or at least make certain portions in accessible.

We also need a way to evaluate how traversable our generated scenes are. We sort of already have his for Mario with the full simulation A* agent, but that is expensive to run (and also is not perfect). If we're not going to be perfect anyway, we may as well rely on something a bit cheaper to run.

We actually already have code for this in MM-NEAT that we should port to this repo (from Java to Python).

A* is a general algorithm, but the major expense comes from calculating the states in the state space. It can run much quicker using a simplified state representation. Look at these files:
https://github.com/schrum2/MM-NEAT/blob/master/src/main/java/edu/southwestern/tasks/mario/level/MarioState.java
https://github.com/schrum2/MM-NEAT/blob/master/src/main/java/edu/southwestern/tasks/megaman/astar/MegaManState.java

Here is the general A* search algorithm that makes use of these state representations:
https://github.com/schrum2/MM-NEAT/blob/master/src/main/java/edu/southwestern/util/search/AStarSearch.java

Port this code into the MarioDiffusion repo. Put most of it in a directory named "astar"

Once this is in place, we can use it in various ways. We need to have some code somewhere that takes a level representation and applies A Star to the level (or scene). Both a starting location and goal location need to be provided in order for this to make sense. This will allow us to assess whether a level is beatable. We should apply this to training data and also to model generated scenes. There are lots of possibilities.

**Comment by schrum2** [comment]
Please find the state representation for Lode Runner in MM-NEAT and port that over as well.

**Comment by fwilson12** [comment]
Will also investigate path visualization tools for interactive tile level gen and ascii data browser

**Comment by schrum2** [comment]
Whenever you get around to testing Mega Man levels, some work done by Ben Capps on this years ago should save you a lot of wasted time:
https://github.com/schrum2/MM-NEAT/issues/473
He was converting all the VGLC levels into MMLV and trying to make them beatable, but we ran into snags on various issues. Some of them can't be solved with our data representation, and it's important to be aware of these limitations during testing.

**Comment by fwilson12** [comment]
The script in the commit above seems to work pretty well, but for megaman, the traversability metric is calculated on whether a state starting on the edge of where the caption defines to be the entrance can reach a tile on the edge of where the caption defines to be the exit; i.e., for a given caption "entrance direction left. exit direction right. ...", traversability is determined by whether any valid state beginning on a tile on the leftmost column can reach a valid tile on the rightmost column. 

However, the caption assignment for megaman can be inconsistent with the entrance and exit descriptors; this is especially noticeable for larger levels that can contain multiple segments that change directions, in which cases the descriptors are rarely accurate. This leads to incorrect traversability assessments as what appears to be a clearly beatable level can be tagged as un-traversable if the exit direction is mislabeled. Not sure what the best way to approach this issue is, but it seems like the next problem to solve in the stack of tasks required to create quality datasets for models to train on.

**Comment by schrum2** [comment]
In the MM-NEAT repo, we mainly applied A* to Mega Man in the context of a complete level rather than individual scenes, and Ben's code would figure out some way to place the exit orb at the end of the final segment heuristically. Look at this thread:
https://github.com/schrum2/MM-NEAT/issues/593
Also look at this code:
https://github.com/schrum2/MM-NEAT/blob/5ca8e9e952dc825d4bf1a44ecd3a8116ca7eb286/src/main/java/edu/southwestern/tasks/megaman/gan/MegaManGANUtil.java#L249
It is responsible for placing the exit orb in the level. Seeing this code, I am very annoyed by all the magic numbers it uses. The tile values are defined here:
https://github.com/schrum2/MM-NEAT/blob/5ca8e9e952dc825d4bf1a44ecd3a8116ca7eb286/src/main/java/edu/southwestern/tasks/megaman/astar/MegaManState.java#L44

The MEGA_MAN_TILE_ORB is 7, which is the important one to notice here.

Anyway, something like this code could be used, but we also need to think in terms of complete levels rather than just segments. I'll share some ideas about that in another issue.

In the meantime, report on how the A* representation does in Mario levels, since the exit condition in those is much easier (do to the right). I would also like to see some visualizations of the paths through the levels.

Don't forget to apply this to Lode Runner as well.

**Comment by fwilson12** [comment]
Ported the visualization tools from MM-NEAT and added a visualization option to the the traversability script that was created in my most recent commit, results so far are good for both mario and megaman; next I'll see how it does on larger megaman scenes and also ensure it works for lode runner.

##

Mario Samples:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/a22ab5a8-d193-496b-a1a6-01f2c44cbc74" />
<img width="2048" height="256" alt="Image" src="https://github.com/user-attachments/assets/f62ad7ba-4319-43cd-a4e4-cbaca86fa925" />
<img width="1024" height="256" alt="Image" src="https://github.com/user-attachments/assets/c7e98183-cc5c-4d49-a825-72a13cda441c" />

##

Megaman 16x14 samples:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/6bf9bc60-2c7c-4dbf-9570-0d33791fa6e0" />

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/b5f6e44b-50e1-4a4c-950b-f9d4689243b2" />

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/6ab1ca91-e810-4bc3-bea0-3e69fbe29eae" />

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/1f8d38e3-49c9-4a2e-80a9-ff9d7339716f" />

**Comment by fwilson12** [comment]
After testing the script with lode runner levels it's clear the game's massive state space makes finding a path very costly, and currently the majority of the levels exceed the visited state budget when being searched by the a* algorithm.  I'll see what I can do to improve it after I work on the following:

For now, I'm focusing on larger megaman scenes, and am working on an algorithm that will place an orb at a suitable place within a segment (code from MM-NEAT is helpful but can't be directly ported as it doesn't generalize to larger scenes)

After this I can implement a similar algorithm for spawn point placement, which should essentially be the same, but the search would start from the other side of a given scene. 

This is mainly for assessing traversability in individual scenes, which can be used to filter un-traversable scenes from training sets. The same algorithms could be applied to a complete level that hopefully we'll be able to generate in the interactive tile level gen, but it'll likely be better to leave picking the spawn and orb up to the user anyway

**Comment by schrum2** [comment]
Incorporate an option to see these visualizations in as many places as possible. In various GUIs you can add a "Simple A*" button that overlays the image with this visualization. This should be added to the ASCII data browser and the interactive tile GUI. There should also be an option with run_diffusion regarding whether or not outputted scenes should include the visualization ... if this is turned on, then the version with the path should be in addition to the original version, not instead of. text_to_level_diffusion.py should ask the user if they want to see the path on their level image.

We might add a similar option elsewhere, but I think this is enough for now.

**Comment by fwilson12** [comment]
Just added basic heuristics like the ones found in MM-NEAT that algorithmically place a spawnpoint and orb in a given MM level, and changed the heuristic from the directional caption-based entrance-to-exit edge heuristic to the more straightforward Chebyshev distance-to-orb heuristic. 

The change was an overall improvement, especially for larger scenes whose faulty captions made the former approach incompatible, but the results still aren't perfect. The heuristic (which scans the level R->L, T->B to place the orb, and L->R, B->T to place a spawn point) sometimes will place strange spawns/orbs, and for vertical level segments that are intended to run from top-to-bottom are typically trivialized by the spawn placement heuristic. 

With that being said, the overall accuracy is better compared to the directional heuristic, even if the spawn/orb is placed weirdly. Below is a variety of examples (un-traversable levels show the spawn in green and orb in blue, but do not feature the red connecting path):

## 32x32

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/0ce37df0-c5e2-433d-9f1d-b7cf76ccd7f5" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/1c2f9c57-43b3-4608-9ce7-90ae825ff2d5" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/8514d7ad-7337-4568-93fa-4edfb474f997" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/21eacb30-d92b-4be0-983d-707d8ceeaa20" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/beb409be-7133-45e4-bcd2-da31bc361635" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/374de504-50ba-4625-81c3-e954d1beaaf4" />

<img width="512" height="512" alt="Image" src="https://github.com/user-attachments/assets/82b9e6c3-5364-4cf3-b0ab-9a8db2fa763d" />

## 64x64

<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/695e494e-30f1-435a-9b06-124ce00ad7b2" />

<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/f24871b8-c867-476c-bd01-9c521b86c2b0" />

<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/9b65ca18-8f8b-4039-bfae-6a2948760cf4" />

(the agent below got so close but it seems the current state renders the last jump impossible)
<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/63ee8891-0e02-4e69-a166-acf7b296e583" />
(interesting note/bug: below, the spawn was placed in a nx1 gap, which means megaman's head was in a block which allowed him to jump through, which actually ended up working better even though the level wasn't traversable anyway)
<img width="1024" height="1024" alt="Image" src="https://github.com/user-attachments/assets/91358e2e-5508-454f-a951-e96077ae2908" />


##

After observing the results from both methods, I think the biggest remaining blocker is still the way we sample larger scenes, as right now it seems unavoidable to not capture a handful of weird, ambiguous scenes. Not sure what the best next steps are; I could try creating a filtered dataset of larger scenes using the assessments from this script and see if models trained on such filtered data might produce more traversable levels.

For now though, I'll work on interfacing the new visualization script with the existing GUIs, along with adding a visualization option as a CLI for run_diffusion and text_to_level_diffusion.

**Comment by schrum2** [comment]
Mega Man seems good enough. I definitely want to be able to override the start/end locations (as described in Issues like #169) by directly placing them at desired locations in designed levels, but the heuristic placement is fine for levels that don't define start/end.

The only thing left before this issue can be closed is some results from Lode Runner.

**Comment by schrum2** [comment]
Correction: integrate the visualization tools into the various GUIs (you were already doing this), add support for Lode Runner, _then_ close. Sorry for any confusion

**Comment by fwilson12** [comment]
GUIs have the simple astar visualization integrated, added an astar visualization as a CLI arg in run/text_to_level diffusion.py, solved the lode runner search space issue
